### DISCUSSION TITLE
@time_trigger and additional params

### DISCUSSION THREAD

**Discussion by stefanuytterhoeven** [custom-components/pyscript#326] [Q&A]
I'm writing a trigger closure, that creates a @time_trigger

def create_triggers(config):
      start_time=config['periode']['start']
       @time_trigger(f"once({start_time})")
            def start_tuinlichten(**kwargs):
                .... turn_on light1, turn_on light2....

but, I also have config['entities'] that contains ... entities :)
I wanted to add that  variable config['entities'] to the @time_trigger(.... as param, so I can use the contens in the function... But HOW?

In de docs, under time trigger , I read:
"Additional keyword parameters can be passed to the trigger function by setting the optional kwargs parameter to a dict with the keywords and values."
But, I don't know how to do this?
Can someone help?

thanks!

**Comment by craigbarratt** [comment]
There are a couple of choices.  Most easily, you can simply use config['entities'] in the trigger function. There's no need to pass it as an argument since it's already in scope (ie, exists) when the trigger closure is defined.

The documentation is referring to how you can pass extra arguments to the trigger function.  That's helpful if you have several different triggers applied to the same function and you'd like to tell them apart.

Here's how that should work. Inside the `@trigger` arguments, assign the variable `kwargs` to a dict with all the parameters you want to pass to the trigger function when that trigger occurs.  Then those `dict` settings will be added as function arguments.  The function can accept those as named parameters (ie, assign to `None` in the declaration), or you can just use `**kwargs` to pick them all up:
```python
@time_trigger("once(sunrise)", kwargs={"var1": 10, "var2": "hello"})
@time_trigger("once(sunset)", kwargs={"var1": 25, "var2": "goodbye"})
def func_sunrise_or_sunset(**kargs):
    ....
```
(note: untested)

**Reply by stefanuytterhoeven** [reply]
Thanks!
Didn't find an example.
Thank you very much

**Reply by stefanuytterhoeven** [reply]
```
registered_triggers = []


def setup_tuinlichten_group(config):
    
    if 'start' in config['periode']:
        notify.persistent_notification(message=f"{config['entities']},{config['periode']['start']}", title="tuinlichtenstart")  
        start_time=config['periode']['start']
        log.info(f"before time trigger {config['entities']}")
        @time_trigger(f"once({start_time})")
        def start_tuinlichten(**kwargs):
            try:
                log.info(f"IN time trigger {config}")
            except:
                log.info("Failed IN time trigger config")
            try:
                log.info(f"IN time trigger {config['entities']}")
            except:
                log.info(f"Failed IN time trigger config['entities']")


    registered_triggers.append(start_tuinlichten)




@time_trigger('startup')      
def tuinlichten_startup():
    log.info("*** start ****")
    for config in pyscript.app_config:
        setup_tuinlichten_group(config)

```

YAML:
```
tuinlichten:
  - entities:
      - light.tuin_hue_lightstrip_outdoor
      - switch.tuin_hue_smart_plug_links
      - switch.tuin_hue_smart_plug_rechts
    periode:
      start: "07:50"
      end: input_datetime.tuinlichten_eindtijd
```

For now, I have only 1 'group' of entities in Yaml. In the future maybe more...
That's why I Use trigger closures. 
When I try this code, I see in the logging:
2022-02-24 07:50:00 INFO (MainThread) [custom_components.pyscript.apps.tuinlichten.start_tuinlichten] Failed IN time trigger config
2022-02-24 07:50:00 INFO (MainThread) [custom_components.pyscript.apps.tuinlichten.start_tuinlichten] Failed IN time trigger config['entities']

I've done this before, with succes, but not with a time_trigger, but state_triggers.... Don't understand it....

update: I added a state_triggert too, and it didn't work either. I have another script, where I have no problem.
There's probably a stupid error, but I don't see it :( 

Update2: FOUND IT (sort of)
I only create the two triggers when "start" is in "config" var
When I remove the IF statement (and adjust the indentation), it works!
Why? No idea? Shouldn't make any difference...

Update3: found a "solution"
this WORKS!

```
registered_triggers = []


def setup_tuinlichten_group(config):
    
    if 'start' in config['periode']:
        notify.persistent_notification(message=f"{config['entities']},{config['periode']['start']}", title="tuinlichtenstart")  
        create_time_trigger(config)
        
        


def create_time_trigger(config):
    start_time=config['periode']['start']
    @time_trigger(f"once({start_time})")
    def start_tuinlichten(**kwargs):
        notify.persistent_notification(message=f"{kwargs}", title="startlicht")  
        log.info(f"globals {globals()}")
        log.info(f"locals {locals()}")
        try:
            log.info(f"IN time trigger {config}")
        except:
            log.info("Failed IN time trigger config")
        try:
            log.info(f"IN time trigger {config['entities']}")
        except:
            log.info(f"Failed IN time trigger config['entities']")

    registered_triggers.append(start_tuinlichten)

@time_trigger('startup')      
def tuinlichten_startup():
    log.info("*** start ****")
    for config in pyscript.app_config:
        setup_tuinlichten_group(config)
```
As you can see, because the "IF" statement seems to cause a problem,
a created a separate function "create_time_trigger(config)" that creates the time trigger...
And this works... Seems very strange to me , but, it works
